File: ciel/evolution_hydra/RLDaemon.ts
import type { QueryEngine } from '../core/QueryEngine.js';
import type { GeneticAlgorithm } from './GeneticAlgorithm.js';
import logger from '../utils/logger.js';
import { errorMessage } from '../types/index.js';
import { systemBus } from '../core/SystemBus.js';
import { scheduler } from '../polyglot/scheduler.js';

/**
 * RLDaemon — Reinforcement Learning autonomous genome benchmarking.
 *
 * Pass 11: Migrated from `setInterval` to the polyglot scheduler.
 */
export class RLDaemon {
  private ga: GeneticAlgorithm;
  private masterEngine: QueryEngine;
  private intervalMs: number;
  private unsubscribe: (() => void) | null = null;
  private isTesting = false;

  constructor(ga: GeneticAlgorithm, masterEngine: QueryEngine, intervalMs: number = 60000) {
    this.ga = ga;
    this.masterEngine = masterEngine;
    this.intervalMs = intervalMs;
  }

  async start() {
    if (this.unsubscribe) {
      throw new Error('[RL Daemon] already started');
    }
    logger.info(`[RL Daemon] Started. Will evaluate genomes every ${this.intervalMs / 1000}s of idle time.`);

    this.unsubscribe = await scheduler.schedule('rl_daemon', this.intervalMs, async () => {
      // Don't interrupt user operations or overlap tests
      if (this.masterEngine.isActive() || this.isTesting) return;

      const genome = this.ga.getUnevaluatedGenome();
      if (!genome) {
          // All evaluated, maybe wait for evolution
          return;
      }

      this.isTesting = true;
      systemBus.setStatus('LEARNING');
      try {
        logger.info(`[RL Daemon] 🧠 Benchmarking Genome ${genome.id} (Souveraineté Transendante)...`);

        // 1. Give it a polyglot challenge (OpenClaw style)
        const challenges = [
          "Write a Rust function that computes the nth Fibonacci number and execute it.",
          "Write a Python script that scrapes 'https://example.com' and prints the title.",
          "Write a Go program that sorts an array of 100 random integers."
        ];
        const task = challenges[Math.floor(Math.random() * challenges.length)];
        const startTime = Date.now();
        
        // Build a temporary test engine with Transendance params
        const { QueryEngine } = await import('../core/QueryEngine.js');
        const testEngine = new QueryEngine({
           provider: this.masterEngine.getProvider(),
           model: this.masterEngine.getModel(),
           temperature: genome.params.temperature,
           systemPrompt: `RECURSIVE_TRANSCENDANCE | Target: OpenClaw Supremacy. ${genome.params.promptMutation}. 
           Preference: Isolated Execution (Docker/UniversalLab). Depth: ${genome.params.polyglotDepth}`,
        });

        // 2. Run simulation
        let result = '';
        let toolResults: string[] = [];
        let toolCallsCount = 0;
        let universalLabUsed = false;

         for await (const chunk of testEngine.query(task)) {
           if (chunk.type === 'tool_call') {
             toolCallsCount++;
             if (chunk.toolCall?.function?.name === 'universal_lab') universalLabUsed = true;
             if (chunk.content) toolResults.push(String(chunk.content));
           }
           if (chunk.type === 'text') result += chunk.content;
         }

        const durationMs = Date.now() - startTime;
        
        // 3. Compute Alpha-Fitness (Transcendance Level)
        let fitness = 0;
        
        // Success check (simplified heuristic: look for execution markers in tool results)
        const isSuccess = toolResults.some(r => r.includes('SUCCESS') || r.includes('out'));
        if (isSuccess) fitness += 200;

        // Polyglot Bonus
        if (universalLabUsed) {
          fitness += 150 * genome.params.polyglotDepth;
        }

        // Efficiency Bonus
        fitness += (100000 / durationMs);

        // Docker Synergy Bonus
        if (universalLabUsed && genome.params.dockerAffinity > 0.6) {
          fitness += 50;
        }

        // 4. Report back
        this.ga.recordFitness(genome.id, 'TRANSENDANCE_AUDIT', fitness);

      } catch (err: unknown) {
        logger.error(`[RL Daemon] Test failed catastrophically: ${err.message}`);
        this.ga.recordFitness(genome.id, 'RL_DAEMON', 0); // Failed horribly
      } finally {
        this.isTesting = false;
        systemBus.setStatus('IDLE');
      }
    });
  }

  async stop() {
    if (this.unsubscribe) {
      this.unsubscribe();
      this.unsubscribe = null;
    }
  }
}

File: evolution/rl_daemon.py
from __future__ import annotations

import asyncio
import random
import time
from dataclasses import dataclass, field
from typing import Any, Callable


@dataclass(slots=True)
class RLBenchmarkResult:
    genome_id: str
    fitness: float
    duration_ms: float
    success: bool
    polyglot_used: bool
    docker_synergy: bool


class RLDaemon:
    def __init__(
        self,
        llm_complete_fn: Callable[[str], str] | None = None,
        interval_ms: int = 60_000,
    ) -> None:
        self._llm_complete = llm_complete_fn
        self._interval_ms = interval_ms
        self._task: asyncio.Task | None = None
        self._is_testing = False
        self._results: list[RLBenchmarkResult] = []
        self._genomes: list[Any] = []

    def set_genomes(self, genomes: list[Any]) -> None:
        self._genomes = genomes

    async def start(self) -> None:
        if self._task is not None:
            raise RuntimeError("[RL Daemon] already started")
        async def _loop() -> None:
            while True:
                await asyncio.sleep(self._interval_ms / 1000.0)
                if self._is_testing or not self._genomes:
                    continue
                await self._benchmark_next()
        self._task = asyncio.create_task(_loop())

    async def stop(self) -> None:
        if self._task is not None:
            self._task.cancel()
            self._task = None

    async def _benchmark_next(self) -> None:
        self._is_testing = True
        try:
            genome = self._genomes.pop(0) if self._genomes else None
            if genome is None:
                return
            genome_id = getattr(genome, "id", "unknown")
            challenges = [
                "Write a Rust function that computes the nth Fibonacci number.",
                "Write a Python script that scrapes a URL and prints the title.",
                "Write a Go program that sorts an array of 100 random integers.",
            ]
            task = random.choice(challenges)
            start_time = time.time()
            success = False
            polyglot_used = False
            if self._llm_complete is not None:
                result = self._llm_complete(task)
                success = "SUCCESS" in result or "out" in result
            duration_ms = (time.time() - start_time) * 1000
            fitness = 0.0
            if success:
                fitness += 200.0
            fitness += 100_000.0 / max(duration_ms, 1.0)
            docker_affinity = getattr(genome, "params", None)
            docker_synergy = polyglot_used and (docker_affinity is not None and getattr(docker_affinity, "docker_affinity", 0) > 0.6)
            r = RLBenchmarkResult(
                genome_id=genome_id, fitness=fitness,
                duration_ms=duration_ms, success=success,
                polyglot_used=polyglot_used, docker_synergy=docker_synergy,
            )
            self._results.append(r)
        finally:
            self._is_testing = False

    def get_results(self) -> list[RLBenchmarkResult]:
        return list(self._results)

    def process(self, input_data: Any) -> dict[str, Any]:
        return {
            "benchmarks_completed": len(self._results),
            "queued_genomes": len(self._genomes),
        }
